src/sparse/KokkosSparse_BlockCrsMatrix.hpp

```diff
@@ -0,0 +1,940 @@
+                  const OrdinalType blockDim)
+  {
+    typedef typename KokkosSparse::CrsMatrix<SType, OType, DType, MTType, IType> crs_matrix_type;
+    typedef typename crs_matrix_type::StaticCrsGraphType crs_graph_type;
+    typedef typename crs_graph_type::entries_type crs_graph_entries_type;
+    typedef typename crs_graph_type::row_map_type crs_graph_row_map_type;
+
+    blockDim_ = blockDim;
+    numCols_ = crs_mtx.numCols() / blockDim;
+    values = crs_mtx.values;
+
+    OrdinalType nbrows = crs_mtx.numRows()/blockDim; // actual number of block rows; add 1 for ptr length
+
+    // block_rows will accumulate the number of blocks per row - this is NOT the row_map with cum sum!!
+    std::vector<OrdinalType> block_rows( nbrows, 0 );
+
+    typename crs_graph_row_map_type::HostMirror h_crs_row_map = Kokkos::create_mirror_view(crs_mtx.graph.row_map);
+    Kokkos::deep_copy(h_crs_row_map, crs_mtx.graph.row_map);
+    typename crs_graph_entries_type::HostMirror h_crs_entries = Kokkos::create_mirror_view(crs_mtx.graph.entries);
+    Kokkos::deep_copy(h_crs_entries, crs_mtx.graph.entries);
+
+    // determine size of block cols indices == number of blocks, i.e. nnz for the block CRS graph
+    OrdinalType numBlocks = 0;
+    for ( OrdinalType i = 0; i < crs_mtx.numRows(); i+=blockDim ) {
+      numBlocks += ( h_crs_row_map(i+1) - h_crs_row_map(i) ) / blockDim; // cum sum
+      // block_rows[ i/blockDim + 1 ] = numBlocks; // cum sum
+      block_rows[ i/blockDim ] = ( h_crs_row_map(i+1) - h_crs_row_map(i) ) / blockDim; // frequency counts
+    }
+
+    // create_staticcrsgraph takes the frequency of blocks per row and returns the cum sum pointer row_map with nbrows+1 size, and total numBlocks in the final entry
+    graph = Kokkos::create_staticcrsgraph<StaticCrsGraphType> ("blockgraph", block_rows);
+    typename values_type::HostMirror h_values = Kokkos::create_mirror_view (values);
+    typename index_type::HostMirror h_entries = Kokkos::create_mirror_view (graph.entries);
+
+    for (OrdinalType i = 0; i < nbrows; ++i) {
+      OrdinalType blks_in_row = block_rows[i];
+      
+      // graph.row_map(i) returns the number of blocks until the current block-row i; 
+      // alternatively the offset start into the block colidx
+      OrdinalType offset_into_blkcolidx_start = graph.row_map(i);
+      OrdinalType offset_into_colidx_start = offset_into_blkcolidx_start*blockDim*blockDim;
+
+      for ( OrdinalType lidx = 0; lidx < blks_in_row; ++lidx ) {
+        h_entries( offset_into_blkcolidx_start+lidx ) = h_crs_entries( offset_into_colidx_start + blockDim*lidx ) / blockDim;
+      }
+    }
+
+    Kokkos::deep_copy (graph.entries, h_entries);
+
+#ifdef KOKKOS_USE_CUSPARSE
+    cusparse_handle (crs_mtx.cusparse_handle),
+    cusparse_descr (crs_mtx.cusparse_descr),
+#endif // KOKKOS_USE_CUSPARSE
+  }
+
+
+  /// Declaration for import - this member function is not inlined
+  void
```

Comment:
It's confusing that this function is called "import"; see above comment.

Reply:
Yes, responded to comment below to see if this can be removed (or at least renamed and hidden).